File: codes/lua/encapsulation/person.lua
local Person = {}
Person.__index = Person

function Person.new(name)    -- The constructor
  local self = {}
  local _name = name;
  self.luckyNumber = 1;
  self.info1 = function()
    print("My name: " .. _name)
  end
  return setmetatable(self, Person)
end

function Person:info()
  self.info1()
  print("My lucky number: " .. self.luckyNumber)
end

return Person

File: codes/ruby/encapsulation/person.rb
class Person

    @@nbr = 0

    attr_accessor :luckyNumber #attr_reader, attr_writer
    attr_reader :t, :name
    protected :name

    def initialize(name)
        @name = name
        @num = @@nbr
        @t = "person"
        @luckyNumber = 0
        @@nbr += 1
    end

    def info
        puts "My name: #@name"
        info1
        info2
        puts "My luckyNumber is: #@luckyNumber"
    end

    private
    def info1
        puts "My number: #@num"
    end

    protected
    def info2
        puts "I am a:  #@t"
    end

    public
    def copy (other)
        @luckyNumber = other.luckyNumber
        @name = other.name
        #@num = other.num #can not access num of other
    end

    def other_info (other)
        puts "other info"
        #other.info1 #private
        other.info2
    end
end
